client/allocrunner/taskrunner/template/template_test.go

```diff
@@ -1403,27 +1434,44 @@ func TestTaskTemplateManager_BlockedEvents(t *testing.T) {
 	}
 
 	// Check to see we got a correct message
+	// assert that all 0-4 keys are missing
 	require.Len(harness.mockHooks.Events, 1)
+	t.Logf("first message: %v", harness.mockHooks.Events[0])
+	missing, more := missingKeys(harness.mockHooks.Events[0])
+	require.Equal(5, len(missing)+more)
 	require.Contains(harness.mockHooks.Events[0].DisplayMessage, "and 2 more")
 
-	// Write 3 keys to Consul
+	// Write 0-2 keys to Consul
 	for i := 0; i < 3; i++ {
 		harness.consul.SetKV(t, fmt.Sprintf("%d", i), []byte{0xa})
 	}
 
 	// Ensure that we get a blocked event
-	select {
-	case <-harness.mockHooks.UnblockCh:
-		t.Fatalf("Task unblock should have not have been called")
-	case <-harness.mockHooks.EmitEventCh:
-	case <-time.After(time.Duration(1*testutil.TestMultiplier()) * time.Second):
-		t.Fatalf("timeout")
+	isExpectedFinalEvent := func(e *structs.TaskEvent) bool {
+		missing, more := missingKeys(e)
+		return reflect.DeepEqual(missing, []string{"3", "4"}) && more == 0
+	}
+	timeout := time.After(time.Second * time.Duration(testutil.TestMultiplier()))
+WAIT_LOOP:
+	for {
+		select {
+		case <-harness.mockHooks.UnblockCh:
+			t.Errorf("Task unblock should have not have been called")
+		case e := <-harness.mockHooks.EmitEventCh:
+			t.Logf("received event: %v", e.DisplayMessage)
+			if isExpectedFinalEvent(e) {
+				break WAIT_LOOP
+			}
+		case <-timeout:
+			t.Errorf("timeout")
+		}
 	}
 
-	// TODO
 	// Check to see we got a correct message
-	eventMsg := harness.mockHooks.Events[len(harness.mockHooks.Events)-1].DisplayMessage
-	if !strings.Contains(eventMsg, "Missing") || strings.Contains(eventMsg, "more") {
-		t.Fatalf("bad event: %q", eventMsg)
+	event := harness.mockHooks.Events[len(harness.mockHooks.Events)-1]
+	if !isExpectedFinalEvent(event) {
+		t.Logf("received all events: %v", pretty.Sprint(harness.mockHooks.Events))
```

Comment:
nitpick: this log message seems redundant with the `t.Fatalf` below and with the logging of event we received above.

Reply:
There is redundancy - this log line emits all messages so it's easy to see the history of events and all of their data, while the Fatal only shows the last message of the "bad" event without its other metadata.

Reply:
Yeah - there is a bit of redundancy.  This emits all the task event data (and that's a bit big) potentially useful for debugging the issue, while others only emit the display messages.